_posts/2016-04-01-new-post.md

```diff
@@ -0,0 +1,912 @@
+      get(this, 'pollInterval')
+    );
+  },
+
+  loadCommentsFromServer() {
+    get(this, 'ajax')
+      .request(get(this, 'url'))
+      .then((comments) => set(this, 'data', comments))
+      .catch((error) => console.error(error));
+  },
+
+  actions: {
+    handleCommentSubmit(comment) {
+      let url = get(this, 'url');
+
+      get(this, 'ajax')
+        .post(url, { data: JSON.stringify(comment) })
+        .then((comments) => set(this, 'data', comments))
+        .catch((error) => console.error(error));
+    }
+  }
+});
+```
+
+> Now that `CommentBox` has made the callback available to `CommentForm` via the
+> `onCommentSubmit` prop, the `CommentForm` can call the callback when the user
+> submits the form:
+
+TODO: Ember diff
+
+```javascript
+// app/components/comment-form.js
+
+import Ember from 'ember';
+
+const {
+  Component,
+  get,
+  setProperties
+} = Ember;
+
+export default Component.extend({
+  init() {
+    this._super(...arguments);
+
+    this.author = '';
+    this.text = '';
+  },
+
+  actions: {
+    handleAuthorChange(e) {
+      setProperties(this, { author: e.target.value });
+    },
+
+    handleTextChange(e) {
+      setProperties(this, { text: e.target.value });
+    },
+
+    handleSubmit(e) {
+      e.preventDefault();
+
+      let author = get(this, 'author').trim();
+      let text = get(this, 'text').trim();
+
+      if (!author || !text) {
+        return;
+      }
+
+      get(this, 'onCommentSubmit')({ author, text });
+      setProperties(this, { author: '', text: '' });
+    }
+  }
+});
+```
+
+### Optimization: Optimistic updates
+
+> Our application is now feature complete but it feels slow to have to wait for
+> the request to complete before your comment appears in the list. We can
+> optimistically add this comment to the list to make the app feel faster.
+
+```javascript
+// app/comments/comment-box.js
+
+import Ember from 'ember';
+
+const {
+  Component,
+  get,
+  inject,
+  set
+} = Ember;
+
+const { service } = inject;
+
+export default Component.extend({
+  ajax: service(),
+
+  init() {
+    this._super(...arguments);
+
+    this.data = [];
+
+    this.loadCommentsFromServer();
+
+    setInterval(
+      this.loadCommentsFromServer.bind(this),
+      get(this, 'pollInterval')
+    );
+  },
+
+  loadCommentsFromServer() {
+    get(this, 'ajax')
+      .request(get(this, 'url'))
+      .then((comments) => set(this, 'data', comments))
+      .catch((error) => console.error(error));
+  },
+
+  actions: {
+    handleCommentSubmit(comment) {
+      let oldComments = get(this, 'data');
+      let url = get(this, 'url');
+
+      set(comment, 'id', Date.now());
+
+      let newComments = oldComments.concat([comment]);
+      set(this, 'data', newComments);
+
+      get(this, 'ajax')
+        .post(url, { data: JSON.stringify(comment) })
+        .then((comments) => set(this, 'data', comments))
+        .catch((error) => {
+          set(this, 'data', oldComments);
```

Comment:
yes it's easy to make this act weird. I didn't want to change the behavior. It's actually more noticable with this using mirage because there's an artificial delay added to the mocked requests whereas the react one uses a little server that just responds like instantly.

It's also weird because the way this endpoint works is that it sends ALL the comments back in the response to the create, which I've never seen as a real API. I was just trying to make it match the server that the React tutorial provides. You can actually pull that one down and run it instead of mirage and this app works 😄